apps/bfd-shared-utils/src/main/java/gov/cms/bfd/sharedutils/generators/token/parser/AbstractTokenGroupParser.java

```diff
@@ -0,0 +1,78 @@
+package gov.cms.bfd.sharedutils.generators.token.parser;
+
+import gov.cms.bfd.sharedutils.generators.exceptions.ParsingException;
+import gov.cms.bfd.sharedutils.generators.token.TokenParserFactory;
+import gov.cms.bfd.sharedutils.generators.token.pattern.TokenPattern;
+import gov.cms.bfd.sharedutils.generators.token.pattern.TokenRepeat;
+import java.util.Collection;
+import java.util.Queue;
+
+/**
+ * This is the base class for the group parsing implementations.
+ *
+ * @param <C> The type of collection the implementing class uses for it's internal container.
+ */
+public abstract class AbstractTokenGroupParser<C extends Collection<TokenPattern>>
+    implements TokenParser {
+
+  @Override
+  public TokenPattern parse(Queue<Character> patternStream) {
+    C patterns = createCollection();
+
+    TokenPattern previousPattern = null;
+
+    while (!patternStream.isEmpty()) {
+      TokenParser tokenParser = TokenParserFactory.createTokenParser(patternStream);
+      TokenPattern pattern = tokenParser.parse(patternStream);
+
+      if (pattern instanceof TokenRepeat) {
+        if (previousPattern != null) {
+          pattern = new TokenRepeat(previousPattern, (TokenRepeat) pattern);
+          pattern.init();
+          patterns.add(pattern);
+          previousPattern = null;
+        } else {
+          throw new ParsingException("Repeat definition has no associated preceding token");
+        }
+      } else {
+        if (previousPattern != null) {
+          patterns.add(previousPattern);
+        }
```

Comment:
Seeing a downcast here is disconcerting.  How about adding previousPattern as an argument to parse() so the abstract base class doesn't have to do any casts.  The TokenRepeat could embed the previousPattern directly.  The parse() method could return a little bean with the TokenPattern it created and a new value (possibly null) for previousPattern or a flag indicating whether it consumed the previousPattern.

Reply:
Yep, I wasn't happy with my use of `instanceof` in this situation, it was a quick solution that I intended to look back at again but forgot to add a TODO.  I'll look at that again, and your suggestion for a refactor 👍